### DISCUSSION TITLE
Versionable breaking changes

### DISCUSSION THREAD

**Discussion by SupinePandora43** [dotnet/runtime#68601] [Ideas]
Replacing `int Length` with `nuint` or `ulong` is a binary and source breaking change, but it can be fixed.

`XXX` can be replaced with .NET version, or some other versioning

# Requirements

## Assembly

* XXX version

## Language

* A way to describe behaviour on previous XXX versions

```cs
public struct Span<T> {
    // ...
    public nuint Length => _length;
    public version(0) int Length => checked((int) _length); // behaviour for XXX version 0
}
```

## Runtime

* Call methods with respect to XXX version of callee's assembly version

```cs
// XXX is 0
span.Length // replaced with a call to Span<T>.get_Length_version_0() by a runtime
```
```cs
// XXX is 1
span.Length
```

# Cons

* adds performance overhead for old programs

**Comment by huoyaoyuan** [comment]
I did consider similar idea. In fact, IL allows overload by return type. The two versions does not have to be different names.

> adds performance overhead for old programs

Unfortunately, array.Length and span.Length are extremely hot paths. Any performance overhead can hurt a lot. This makes the problem more difficult to handle.
